deals/module/deals.go

```diff
@@ -235,61 +235,125 @@ func (m *Module) Watch(ctx context.Context, proposals []cid.Cid) (<-chan deals.D
-// Records are sorted ascending by activation epoch then timestamp.
-func (m *Module) FinalDealRecords() ([]deals.DealRecord, error) {
-	ret, err := m.store.getFinalDeals()
-	if err != nil {
-		return nil, fmt.Errorf("getting final deals: %v", err)
+// ListStorageDealRecords lists storage deals according to the provided options.
+func (m *Module) ListStorageDealRecords(opts ...deals.ListDealRecordsOption) ([]deals.StorageDealRecord, error) {
+	c := deals.ListDealRecordsConfig{}
+	for _, opt := range opts {
+		opt(&c)
+	}
+	var final []deals.StorageDealRecord
+	if !c.OnlyPending {
+		recs, err := m.store.getFinalDeals()
+		if err != nil {
+			return nil, fmt.Errorf("getting final deals: %v", err)
+		}
+		final = recs
 	}
-	return ret, nil
-}
 
-// PendingDealRecords returns a list of all pending storage deals.
-// Records are sorted ascending by timestamp.
-func (m *Module) PendingDealRecords() ([]deals.DealRecord, error) {
-	ret, err := m.store.getPendingDeals()
-	if err != nil {
-		return nil, fmt.Errorf("getting pending deals: %v", err)
+	var pending []deals.StorageDealRecord
+	if c.IncludePending || c.OnlyPending {
+		recs, err := m.store.getPendingDeals()
+		if err != nil {
+			return nil, fmt.Errorf("getting pending deals: %v", err)
+		}
+		pending = recs
 	}
-	return ret, nil
-}
 
-// AllDealRecords returns a list of all finalized and pending deals.
-// Records are sorted ascending by activation epoch, if available, then timestamp.
-func (m *Module) AllDealRecords() ([]deals.DealRecord, error) {
-	ret, err := m.store.getFinalDeals()
-	if err != nil {
-		return nil, fmt.Errorf("getting final deals: %v", err)
-	}
-	pending, err := m.store.getPendingDeals()
-	if err != nil {
-		return nil, fmt.Errorf("getting pending deals: %v", err)
+	combined := append(final, pending...)
+
+	var filtered []deals.StorageDealRecord
+
+	if len(c.FromAddrs) > 0 || len(c.DataCids) > 0 {
+		var fromAddrsFilter map[string]struct{}
+		var dataCidsFilter map[string]struct{}
+		for _, addr := range c.FromAddrs {
+			fromAddrsFilter[addr] = struct{}{}
+		}
+		for _, cid := range c.DataCids {
+			dataCidsFilter[cid] = struct{}{}
+		}
+		for _, record := range combined {
+			_, inFromAddrsFilter := fromAddrsFilter[record.Addr]
+			_, inDataCidsFilter := dataCidsFilter[record.DealInfo.PieceCID.String()]
+			includeViaFromAddrs := len(c.FromAddrs) == 0 || inFromAddrsFilter
+			includeViaDataCids := len(c.DataCids) == 0 || inDataCidsFilter
+			if includeViaFromAddrs && includeViaDataCids {
+				filtered = append(filtered, record)
+			}
+		}
+	} else {
+		filtered = combined
 	}
-	ret = append(ret, pending...)
-	sort.Slice(ret, func(i, j int) bool {
-		if ret[i].Pending || ret[j].Pending {
-			return ret[i].Time < ret[j].Time
+
+	sort.Slice(filtered, func(i, j int) bool {
+		l := filtered[j]
+		r := filtered[i]
+		if c.Ascending {
+			l = filtered[i]
+			r = filtered[j]
+		}
+		if l.Pending || r.Pending {
+			return l.Time < r.Time
 		}
```

Comment:
I think there's a possible problem here with the sorting condition.

A pending, B final, C final. B.time<A.time<C.time, and B.Activation>C.Activation.
I think you can get different final sortition depending on the order in which the underlying algorithm sorts things.

Reply:
Hmm, trying to think through this. Can you think of how this could be fixed? I guess I was (probably incorrectly) assuming the underlying sorting algorithm would continue comparing pairs of items until things are settled completely. But that was just an assumption.

Reply:
Will be sure to cover your case in a test.

Reply:
The algorithm used by sort depends on the number of elements of the slice.
The important thing with the sorted condition is that it should be transitive, so A<B and B<C should always imply A<C. If that isn't the case, depending on the original slice circumstances, you can get different results.

Maybe just placing Pending first (with Time to subsort), and Finalized afterwards (with activation epoch) should be reasonable. Both deals are quite different and are trying to use unrelated temporal dimensions, so its quite hard to reconcile both at the same time.

Reply:
Or maybe always by `Time`.  I'd let the application decide a more complex sortition criteria depending on what is trying to do?

Reply:
Good call, keep it simple.